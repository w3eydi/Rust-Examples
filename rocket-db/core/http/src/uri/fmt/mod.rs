//! Type safe and URI safe formatting types and traits.

mod uri_display;
mod formatter;
mod from_uri_param;
mod encoding;
mod part;

pub use self::formatter::*;
pub use self::uri_display::*;
pub use self::from_uri_param::*;
pub use self::part::*;

pub(crate) use self::encoding::*;
