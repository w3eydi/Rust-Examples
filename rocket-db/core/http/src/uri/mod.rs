//! Types for URIs and traits for rendering URI components.

#[macro_use]
mod uri;
mod origin;
mod reference;
mod authority;
mod absolute;
mod segments;
mod path_query;
mod asterisk;
mod host;

pub mod error;
pub mod fmt;

#[doc(inline)]
pub use self::error::Error;

pub use self::uri::*;
pub use self::authority::*;
pub use self::origin::*;
pub use self::absolute::*;
pub use self::segments::*;
pub use self::reference::*;
pub use self::path_query::*;
pub use self::asterisk::*;
pub use self::host::*;
