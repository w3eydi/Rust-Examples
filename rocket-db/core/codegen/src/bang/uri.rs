use std::fmt::Display;

use devise::{Result, ext::{SpanDiagnosticExt, quote_respanned}};
use syn::{Expr, Ident, Type, spanned::Spanned};
use proc_macro2::TokenStream;

use crate::http::uri::fmt;
use crate::http_codegen::Optional;
use crate::syn_ext::IdentExt;
use crate::bang::uri_parsing::*;
use crate::attribute::param::Parameter;
use crate::exports::*;
use crate::URI_MACRO_PREFIX;

macro_rules! p {
    (@go $num:expr, $singular:expr, $plural:expr) => (
        if $num == 1 { $singular.into() } else { $plural }
    );

    ("parameter", $n:expr) => (p!(@go $n, "parameter", "parameters"));
    ($n:expr, "was") => (p!(@go $n, "1 was", format!("{} were", $n)));
    ($n:expr, "parameter") => (p!(@go $n, "1 parameter", format!("{} parameters", $n)));
}

pub fn prefix_last_segment(path: &mut syn::Path, prefix: &str) {
    let mut last_seg = path.segments.last_mut().expect("syn::Path has segments");
    last_seg.ident = last_seg.ident.prepend(prefix);
}

pub fn _uri_macro(input: TokenStream) -> Result<TokenStream> {
    let input2: TokenStream = input.clone();
    match syn::parse2::<UriMacro>(input)? {
        UriMacro::Routed(ref mut mac) => {
            prefix_last_segment(&mut mac.route.path, URI_MACRO_PREFIX);
            let path = &mac.route.path;
            Ok(quote!(#path!(#input2)))
        },
        UriMacro::Literal(uri) => Ok(quote!(#uri)),
    }
}

fn extract_exprs(internal: &InternalUriParams) -> Result<(
        impl Iterator<Item = &Expr>,             // path exprs
        impl Iterator<Item = &ArgExpr>,          // query exprs
        impl Iterator<Item = (&Ident, &Type)>,   // types for both path || query
    )>
{
    let route_name = &internal.uri_mac.route.path;
    match internal.validate() {
        Validation::Ok(exprs) => {
            let path_params = internal.dynamic_path_params();
            let path_param_count = path_params.clone().count();
            for expr in exprs.iter().take(path_param_count) {
                if expr.as_expr().is_none() {
                    return Err(expr.span().error("path parameters cannot be ignored"));
                }
            }

            let query_exprs = exprs.clone().into_iter().skip(path_param_count);
            let path_exprs = exprs.into_iter().map(|e| e.unwrap_expr()).take(path_param_count);
            let types = internal.fn_args.iter().map(|a| (&a.ident, &a.ty));
            Ok((path_exprs, query_exprs, types))
        }
        Validation::NamedIgnored(_) => {
            let mut route_name = quote!(#route_name).to_string();
            route_name.retain(|c| !c.is_whitespace());

            let diag = internal.uri_mac.args_span()
                .error("expected unnamed arguments due to ignored parameters")
                .note(format!("uri for route `{}` ignores path parameters: \"{}\"",
                        route_name, internal.route_uri));

            Err(diag)
        }
        Validation::Unnamed(expected, actual) => {
            let mut route_name = quote!(#route_name).to_string();
            route_name.retain(|c| !c.is_whitespace());

            let diag = internal.uri_mac.args_span()
                .error(format!("route expects {} but {} supplied",
                         p!(expected, "parameter"), p!(actual, "was")))
                .note(format!("route `{}` has uri \"{}\"", route_name, internal.route_uri));

            Err(diag)
        }
        Validation::Named(missing, extra, dup) => {
            let e = format!("invalid parameters for `{}` route uri", quote!(#route_name));
            let mut diag = internal.uri_mac.args_span().error(e)
                .note(format!("uri parameters are: {}", internal.fn_args_str()));

            fn join<S: Display, T: Iterator<Item = S>>(iter: T) -> (&'static str, String) {
                let mut items: Vec<_> = iter.map(|i| format!("`{}`", i)).collect();
                items.dedup();
                (p!("parameter", items.len()), items.join(", "))
            }

            if !missing.is_empty() {
                let (ps, msg) = join(missing.iter());
                diag = diag.help(format!("missing {}: {}", ps, msg));
            }

            if !extra.is_empty() {
                let (ps, msg) = join(extra.iter());
                let spans: Vec<_> = extra.iter().map(|ident| ident.span()).collect();
                diag = diag.span_help(spans, format!("unknown {}: {}", ps, msg));
            }

            if !dup.is_empty() {
                let (ps, msg) = join(dup.iter());
                let spans: Vec<_> = dup.iter().map(|ident| ident.span()).collect();
                diag = diag.span_help(spans, format!("duplicate {}: {}", ps, msg));
            }

            Err(diag)
        }
    }
}

fn add_binding<P: fmt::Part>(to: &mut Vec<TokenStream>, ident: &Ident, ty: &Type, expr: &Expr) {
    let span = expr.span();
    let part = match P::KIND {
        fmt::Kind::Path => quote_spanned!(span => #_fmt::Path),
        fmt::Kind::Query  => quote_spanned!(span => #_fmt::Query),
    };

    let tmp_ident = ident.clone().with_span(expr.span());
    let let_stmt = quote_spanned!(span => let #tmp_ident = #expr);

    to.push(quote_spanned!(span =>
        #[allow(non_snake_case)] #let_stmt;
        let #ident = <#ty as #_fmt::FromUriParam<#part, _>>::from_uri_param(#tmp_ident);
    ));
}

fn explode_path<'a>(
    internal: &InternalUriParams,
    bindings: &mut Vec<TokenStream>,
    mut exprs: impl Iterator<Item = &'a Expr>,
    mut args: impl Iterator<Item = (&'a Ident, &'a Type)>,
) -> TokenStream {
    if internal.dynamic_path_params().count() == 0 {
        let path = internal.route_uri.path().as_str();
        quote!(#_fmt::UriArgumentsKind::Static(#path))
    } else {
        let uri_display = quote!(#_fmt::UriDisplay<#_fmt::Path>);
        let dyn_exprs = internal.path_params.iter().map(|param| {
            match param {
                Parameter::Static(name) => {
                    quote!(&#name as &dyn #uri_display)
                },
                Parameter::Dynamic(_) | Parameter::Guard(_) => {
                    let (ident, ty) = args.next().expect("ident/ty for non-ignored");
                    let expr = exprs.next().expect("one expr per dynamic arg");
                    add_binding::<fmt::Path>(bindings, &ident, &ty, &expr);
                    quote_spanned!(expr.span() => &#ident as &dyn #uri_display)
                }
                Parameter::Ignored(_) => {
                    let expr = exprs.next().expect("one expr per dynamic arg");
                    quote_spanned!(expr.span() => &#expr as &dyn #uri_display)
                }
            }
        });

        quote!(#_fmt::UriArgumentsKind::Dynamic(&[#(#dyn_exprs),*]))
    }
}

fn explode_query<'a>(
    internal: &InternalUriParams,
    bindings: &mut Vec<TokenStream>,
    mut arg_exprs: impl Iterator<Item = &'a ArgExpr>,
    mut args: impl Iterator<Item = (&'a Ident, &'a Type)>,
) -> Option<TokenStream> {
    let query = internal.route_uri.query()?.as_str();
    if internal.dynamic_query_params().count() == 0 {
        return Some(quote!(#_fmt::UriArgumentsKind::Static(#query)));
    }

    let query_arg = quote!(#_fmt::UriQueryArgument);
    let uri_display = quote!(#_fmt::UriDisplay<#_fmt::Query>);
    let dyn_exprs = internal.query_params.iter().filter_map(|param| {
        if let Parameter::Static(source) = param {
            return Some(quote!(#query_arg::Raw(#source)));
        }

        let dynamic = match param {
            Parameter::Static(source) =>  {
                return Some(quote!(#query_arg::Raw(#source)));
            },
            Parameter::Dynamic(ref seg) => seg,
            Parameter::Guard(ref seg) => seg,
            Parameter::Ignored(_) => unreachable!("invariant: unignorable q")
        };

        let (ident, ty) = args.next().expect("ident/ty for query");
        let arg_expr = arg_exprs.next().expect("one expr per query");
        let expr = match arg_expr.as_expr() {
            Some(expr) => expr,
            None => {
                // Force a typecheck for the `Ignoreable` trait.
                bindings.push(quote_respanned! { arg_expr.span() =>
                    #_fmt::assert_ignorable::<#_fmt::Query, #ty>();
                });

                return None;
            }
        };

        let name = &dynamic.name;
        add_binding::<fmt::Query>(bindings, &ident, &ty, &expr);
        Some(match dynamic.trailing {
            false => quote_spanned! { expr.span() =>
                #query_arg::NameValue(#name, &#ident as &dyn #uri_display)
            },
            true => quote_spanned! { expr.span() =>
                #query_arg::Value(&#ident as &dyn #uri_display)
            },
        })
    });

    Some(quote!(#_fmt::UriArgumentsKind::Dynamic(&[#(#dyn_exprs),*])))
}

pub fn _uri_internal_macro(input: TokenStream) -> Result<TokenStream> {
    // Parse the internal invocation and the user's URI param expressions.
    let internal = syn::parse2::<InternalUriParams>(input)?;
    let (path_exprs, query_exprs, mut fn_args) = extract_exprs(&internal)?;

    let mut bindings = vec![];
    let path = explode_path(&internal, &mut bindings, path_exprs, &mut fn_args);
    let query = Optional(explode_query(&internal, &mut bindings, query_exprs, fn_args));

    let prefix = internal.uri_mac.prefix.as_ref()
        .map(|prefix| quote_spanned!(prefix.span() => .with_prefix(#prefix)));

    let suffix = internal.uri_mac.suffix.as_ref()
        .map(|suffix| quote_spanned!(suffix.span() => .with_suffix(#suffix)));

     Ok(quote_spanned!(internal.uri_mac.route.path.span() =>
         #[allow(unused_braces)] {
             #(#bindings)*
             let __builder = #_fmt::RouteUriBuilder::new(#path, #query);
             __builder #prefix #suffix .render()
         }
     ))
}
