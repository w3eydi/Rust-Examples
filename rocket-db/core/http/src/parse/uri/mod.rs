mod parser;
mod error;
pub(crate) mod tables;

#[cfg(test)] mod tests;

use crate::uri::{Uri, Origin, Absolute, Authority, Reference, Asterisk};

use self::parser::*;

pub use self::error::Error;

type RawInput<'a> = pear::input::Pear<pear::input::Cursor<&'a [u8]>>;

#[inline]
pub fn from_str(s: &str) -> Result<Uri<'_>, Error<'_>> {
    Ok(parse!(uri: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn origin_from_str(s: &str) -> Result<Origin<'_>, Error<'_>> {
    Ok(parse!(origin: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn authority_from_str(s: &str) -> Result<Authority<'_>, Error<'_>> {
    Ok(parse!(authority: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn authority_from_bytes(s: &[u8]) -> Result<Authority<'_>, Error<'_>> {
    Ok(parse!(authority: RawInput::new(s))?)
}

#[inline]
pub fn scheme_from_str(s: &str) -> Result<&str, Error<'_>> {
    let _validated = parse!(scheme: RawInput::new(s.as_bytes()))?;
    Ok(s)
}

#[inline]
pub fn absolute_from_str(s: &str) -> Result<Absolute<'_>, Error<'_>> {
    Ok(parse!(absolute: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn asterisk_from_str(s: &str) -> Result<Asterisk, Error<'_>> {
    Ok(parse!(asterisk: RawInput::new(s.as_bytes()))?)
}

#[inline]
pub fn reference_from_str(s: &str) -> Result<Reference<'_>, Error<'_>> {
    Ok(parse!(reference: RawInput::new(s.as_bytes()))?)
}
