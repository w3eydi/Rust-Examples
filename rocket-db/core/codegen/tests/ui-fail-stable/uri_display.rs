../ui-fail/uri_display.rs