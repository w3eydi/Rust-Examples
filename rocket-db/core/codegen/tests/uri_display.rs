#[macro_use] extern crate rocket;

use rocket::http::uri::fmt::{UriDisplay, Query, Path};
use rocket::serde::{Serialize, Deserialize};

macro_rules! assert_uri_display_query {
    ($v:expr, $expected:expr) => (
        let uri_string = format!("{}", &$v as &dyn UriDisplay<Query>);
        assert_eq!(uri_string, $expected);
    )
}

macro_rules! assert_query_form_roundtrip {
    ($T:ty, $v:expr) => ({
        use rocket::form::{Form, Strict};
        use rocket::http::RawStr;

        let v = $v;
        let string = format!("{}", &v as &dyn UriDisplay<Query>);
        let raw = RawStr::new(&string);
        let value = Form::<Strict<$T>>::parse_encoded(raw).map(|s| s.into_inner());
        assert_eq!(value.expect("form parse"), v);
    })
}

macro_rules! assert_query_value_roundtrip {
    ($T:ty, $v:expr) => ({
        use rocket::form::{Form, Strict};
        use rocket::http::RawStr;

        let v = $v;
        let string = format!("={}", &v as &dyn UriDisplay<Query>);
        let raw = RawStr::new(&string);
        let value = Form::<Strict<$T>>::parse_encoded(raw).map(|s| s.into_inner());
        assert_eq!(value.expect("form parse"), v);
    })
}

#[derive(UriDisplayQuery, Clone)]
enum Foo<'r> {
    First(&'r str),
    Second {
        inner: &'r str,
        other: usize,
    },
    Third {
        #[field(name = "type")]
        kind: String,
    },
}

#[test]
fn uri_display_foo() {
    let foo = Foo::First("hello".into());
    assert_uri_display_query!(foo, "hello");

    let foo = Foo::First("hello there".into());
    assert_uri_display_query!(foo, "hello%20there");

    let foo = Foo::Second { inner: "hi".into(), other: 123 };
    assert_uri_display_query!(foo, "inner=hi&other=123");

    let foo = Foo::Second { inner: "hi bo".into(), other: 321 };
    assert_uri_display_query!(foo, "inner=hi%20bo&other=321");

    let foo = Foo::Third { kind: "hello".into() };
    assert_uri_display_query!(foo, "type=hello");

    let foo = Foo::Third { kind: "hello there".into() };
    assert_uri_display_query!(foo, "type=hello%20there");
}

#[derive(UriDisplayQuery)]
struct Bar<'a> {
    foo: Foo<'a>,
    baz: String,
}

#[test]
fn uri_display_bar() {
    let foo = Foo::First("hello".into());
    let bar = Bar { foo, baz: "well, hi!".into() };
    assert_uri_display_query!(bar, "foo=hello&baz=well,%20hi!");

    let foo = Foo::Second { inner: "hi".into(), other: 123 };
    let bar = Bar { foo, baz: "done".into() };
    assert_uri_display_query!(bar, "foo.inner=hi&foo.other=123&baz=done");

    let foo = Foo::Third { kind: "hello".into() };
    let bar = Bar { foo, baz: "turkey day".into() };
    assert_uri_display_query!(bar, "foo.type=hello&baz=turkey%20day");
}

#[derive(UriDisplayQuery)]
struct Baz<'a> {
    foo: Foo<'a>,
    bar: Bar<'a>,
    last: String
}

#[test]
fn uri_display_baz() {
    let foo1 = Foo::Second { inner: "hi".into(), other: 123 };
    let foo2 = Foo::Second { inner: "bye".into(), other: 321 };
    let bar = Bar { foo: foo2, baz: "done".into() };
    let baz = Baz { foo: foo1, bar, last: "ok".into() };
    assert_uri_display_query!(baz, "foo.inner=hi&foo.other=123&\
                              bar.foo.inner=bye&bar.foo.other=321&bar.baz=done&\
                              last=ok");

    let foo1 = Foo::Third { kind: "hello".into() };
    let foo2 = Foo::First("bye".into());
    let bar = Bar { foo: foo1, baz: "end".into() };
    let baz = Baz { foo: foo2, bar, last: "done".into() };
    assert_uri_display_query!(baz, "foo=bye&\
                              bar.foo.type=hello&bar.baz=end&\
                              last=done");
}

#[derive(UriDisplayQuery)]
struct Bam<'a> {
    foo: &'a str,
    bar: Option<usize>,
    baz: Result<&'a str, usize>,
}

#[test]
fn uri_display_bam() {
    let bam = Bam { foo: "hi hi", bar: Some(1), baz: Err(2) };
    assert_uri_display_query!(bam, "foo=hi%20hi&bar=1");

    let bam = Bam { foo: "hi hi", bar: None, baz: Err(2) };
    assert_uri_display_query!(bam, "foo=hi%20hi");

    let bam = Bam { foo: "hi hi", bar: Some(1), baz: Ok("tony".into()) };
    assert_uri_display_query!(bam, "foo=hi%20hi&bar=1&baz=tony");

    let bam = Bam { foo: "hi hi", bar: None, baz: Ok("tony".into()) };
    assert_uri_display_query!(bam, "foo=hi%20hi&baz=tony");
}

#[test]
fn uri_display_c_like() {
    #[derive(UriDisplayQuery)]
    enum CLike { A, B, C }

    assert_uri_display_query!(CLike::A, "A");
    assert_uri_display_query!(CLike::B, "B");
    assert_uri_display_query!(CLike::C, "C");

    #[derive(UriDisplayQuery)]
    enum CLikeV {
        #[field(value = "a")]
        A,
        #[field(value = "tomato")]
        #[field(value = "juice")]
        B,
        #[field(value = "carrot")]
        C
    }

    assert_uri_display_query!(CLikeV::A, "a");
    assert_uri_display_query!(CLikeV::B, "tomato");
    assert_uri_display_query!(CLikeV::C, "carrot");

    #[derive(UriDisplayQuery)]
    #[allow(non_camel_case_types)]
    enum CLikeR { r#for, r#type, r#async, #[field(value = "stop")] r#yield }

    assert_uri_display_query!(CLikeR::r#for, "for");
    assert_uri_display_query!(CLikeR::r#type, "type");
    assert_uri_display_query!(CLikeR::r#async, "async");
    assert_uri_display_query!(CLikeR::r#yield, "stop");

    #[derive(UriDisplayQuery)]
    struct Nested {
        foo: CLike,
        bar: CLikeV,
        last: CLikeR
    }

    let nested = Nested {
        foo: CLike::B,
        bar: CLikeV::B,
        last: CLikeR::r#type,
    };

    assert_uri_display_query!(nested, "foo=B&bar=tomato&last=type");
}

macro_rules! assert_uri_display_path {
    ($v:expr, $s:expr) => (
        let uri_string = format!("{}", &$v as &dyn UriDisplay<Path>);
        assert_eq!(uri_string, $s);
    )
}

#[derive(UriDisplayPath)]
struct FooP(&'static str);

#[derive(UriDisplayPath)]
struct BarP<'a>(&'a str);

#[derive(UriDisplayPath)]
struct BazP<'a, T>(&'a T);

#[derive(UriDisplayPath)]
struct BamP<T>(T);

#[derive(UriDisplayPath)]
struct BopP(FooP);

#[test]
fn uri_display_path() {
    assert_uri_display_path!(FooP("hi"), "hi");
    assert_uri_display_path!(FooP("hi there"), "hi%20there");
    assert_uri_display_path!(BarP("hi there"), "hi%20there");
    assert_uri_display_path!(BazP(&FooP("hi")), "hi");
    assert_uri_display_path!(BazP(&BarP("hi there")), "hi%20there");
    assert_uri_display_path!(BamP(12), "12");
    assert_uri_display_path!(BamP(BazP(&100)), "100");
    assert_uri_display_path!(BopP(FooP("bop foo")), "bop%20foo");
}

#[test]
fn uri_display_serde() {
    use rocket::serde::json::Json;

    #[derive(Debug, PartialEq, Clone, FromForm, UriDisplayQuery, Deserialize, Serialize)]
    #[serde(crate = "rocket::serde")]
    struct Bam {
        foo: String,
        bar: Option<usize>,
        baz: bool,
    }

    #[derive(Debug, PartialEq, FromForm, UriDisplayQuery)]
    struct JsonFoo(Json<Bam>);

    let bam = Bam {
        foo: "hi[]=there.baz !?".into(),
        bar: None,
        baz: true,
    };

    assert_query_form_roundtrip!(Bam, bam.clone());

    assert_query_value_roundtrip!(JsonFoo, JsonFoo(Json(bam.clone())));

    // FIXME: https://github.com/rust-lang/rust/issues/86706
    #[allow(private_in_public)]
    #[derive(Debug, PartialEq, Clone, FromForm, UriDisplayQuery)]
    struct Q<T>(Json<T>);

    #[derive(Debug, PartialEq, Clone, FromForm, UriDisplayQuery)]
    pub struct Generic<A, B> {
        a: Q<A>,
        b: Q<B>,
        c: Q<A>,
    }

    assert_query_form_roundtrip!(Generic<usize, String>, Generic {
        a: Q(Json(133)),
        b: Q(Json("hello, world#rocket!".into())),
        c: Q(Json(40486)),
    });

    #[derive(Debug, PartialEq, Clone, FromForm, UriDisplayQuery)]
    // This is here to ensure we don't warn, which we can't test with trybuild.
    pub struct GenericBorrow<'a, A: ?Sized, B: 'a> {
        a: Q<&'a A>,
        b: Q<B>,
        c: Q<&'a A>,
    }

    // TODO: This requires `MsgPack` to parse from value form fields.
    //
    // use rocket::serde::msgpack::MsgPack;
    //
    // #[derive(Debug, PartialEq, FromForm, UriDisplayQuery)]
    // struct MsgPackFoo(MsgPack<Bam>);
    //
    // assert_query_value_roundtrip!(MsgPackFoo, MsgPackFoo(MsgPack(bam)));
}
