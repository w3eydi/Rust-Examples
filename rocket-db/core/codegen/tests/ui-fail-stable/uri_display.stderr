error: fieldless structs are not supported
 --> $DIR/uri_display.rs:4:1
  |
4 | struct Foo1;
  | ^^^^^^

error: [note] error occurred while deriving `UriDisplay`
 --> $DIR/uri_display.rs:3:10
  |
3 | #[derive(UriDisplayQuery)]
  |          ^^^^^^^^^^^^^^^
  |
  = note: this error originates in the derive macro `UriDisplayQuery` (in Nightly builds, run with -Z macro-backtrace for more info)

error: fieldless structs are not supported
 --> $DIR/uri_display.rs:7:1
  |
7 | struct Foo2();
  | ^^^^^^

error: [note] error occurred while deriving `UriDisplay`
 --> $DIR/uri_display.rs:6:10
  |
6 | #[derive(UriDisplayQuery)]
  |          ^^^^^^^^^^^^^^^
  |
  = note: this error originates in the derive macro `UriDisplayQuery` (in Nightly builds, run with -Z macro-backtrace for more info)

error: empty enums are not supported
  --> $DIR/uri_display.rs:10:11
   |
10 | enum Foo3 { }
   |           ^^^

error: [note] error occurred while deriving `UriDisplay`
 --> $DIR/uri_display.rs:9:10
  |
9 | #[derive(UriDisplayQuery)]
  |          ^^^^^^^^^^^^^^^
  |
  = note: this error originates in the derive macro `UriDisplayQuery` (in Nightly builds, run with -Z macro-backtrace for more info)

error: tuple structs or variants must have exactly one field
  --> $DIR/uri_display.rs:18:12
   |
18 | struct Foo5(String, String);
   |            ^^^^^^^^^^^^^^^^

error: [note] error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:17:10
   |
17 | #[derive(UriDisplayQuery)]
   |          ^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `UriDisplayQuery` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected list `#[attr(..)]`, found bare integer literal
  --> $DIR/uri_display.rs:22:20
   |
22 |     #[field(name = 123)]
   |                    ^^^

error: [note] error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:20:10
   |
20 | #[derive(UriDisplayQuery)]
   |          ^^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `UriDisplayQuery` (in Nightly builds, run with -Z macro-backtrace for more info)

error: struct must have exactly one field
  --> $DIR/uri_display.rs:27:12
   |
27 | struct Foo7(String, usize);
   |            ^^^^^^^^^^^^^^^

error: [note] error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:26:10
   |
26 | #[derive(UriDisplayPath)]
   |          ^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `UriDisplayPath` (in Nightly builds, run with -Z macro-backtrace for more info)

error: struct must have exactly one field
  --> $DIR/uri_display.rs:30:1
   |
30 | struct Foo8;
   | ^^^^^^

error: [note] error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:29:10
   |
29 | #[derive(UriDisplayPath)]
   |          ^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `UriDisplayPath` (in Nightly builds, run with -Z macro-backtrace for more info)

error: enums are not supported
  --> $DIR/uri_display.rs:33:1
   |
33 | enum Foo9 {  }
   | ^^^^

error: [note] error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:32:10
   |
32 | #[derive(UriDisplayPath)]
   |          ^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `UriDisplayPath` (in Nightly builds, run with -Z macro-backtrace for more info)

error: named structs are not supported
  --> $DIR/uri_display.rs:36:1
   |
36 | struct Foo10 {
   | ^^^^^^

error: [note] error occurred while deriving `UriDisplay`
  --> $DIR/uri_display.rs:35:10
   |
35 | #[derive(UriDisplayPath)]
   |          ^^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `UriDisplayPath` (in Nightly builds, run with -Z macro-backtrace for more info)
